### PR TITLE
[Support][YamlTraits] Add quoting for keys in textual YAML representation

### PR DESCRIPTION
**Pull request by jasilvanus**
The support library contains helpers to parse and emit YAML documents.

In the textual YAML representation, some strings need to be quoted, e.g.
when containing unprintable characters. There are also cases where quoting
isn't strictly necessary, but prevents issues in some cases, for instance strings
consisting of digits only.

We already have such quoting implemented for YAML values.

This patch applies the same quoting to YAML *keys*.

One affected case is output of control registers in AMDGPU Msgpack metadata,
which are printed in a format like this:

```
   0x2cca (SPI_SHADER_PGM_RSRC1_ES): 42
```

With this patch, the key is quoted:

```
   '0x2cca (SPI_SHADER_PGM_RSRC1_ES)': 42
```

Most test changes come from this pattern.

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
Unit tests in llvm/unittests/Support/YAML*?

**Comment by jasilvanus** [comment]
> Unit tests in llvm/unittests/Support/YAML*?

Good point, thanks. I've added unit tests.

**Comment by arsenm** [comment]
> One affected case is output of control registers in AMDGPU Msgpack metadata, which are printed in a format like this:
> 
> ```
>    0x2cca (SPI_SHADER_PGM_RSRC1_ES): 42
> ```
> 
> With this patch, the key is quoted:
> 
> ```
>    '0x2cca (SPI_SHADER_PGM_RSRC1_ES)': 42
> ```

Could we just drop the space in the printing here?

File: llvm/lib/Support/YAMLTraits.cpp [hunk 1]
```diff
@@ -801,6 +769,45 @@ void Output::output(StringRef s) {
...
 }
 
+void Output::output(StringRef S, QuotingType MustQuote) {
+  if (MustQuote == QuotingType::None) {
+    // Only quote if we must.
+    output(S);
+    return;
+  }
+
+  const char *const Quote = MustQuote == QuotingType::Single ? "'" : "\"";
```
Comment (arsenm) [inline llvm/lib/Support/YAMLTraits.cpp:779]: StringLiteral?

File: llvm/lib/Support/YAMLTraits.cpp [hunk 2]
```diff
@@ -801,6 +769,45 @@ void Output::output(StringRef s) {
...
+  unsigned End = S.size();
+  const char *Base = S.data();
+
+  // When using single-quoted strings, any single quote ' must be doubled to be
+  // escaped.
+  while (j < End) {
+    if (S[j] == '\'') {                   // Escape quotes.
+      output(StringRef(&Base[i], j - i)); // "flush".
+      output(StringLiteral("''"));        // Print it as ''
+      i = j + 1;
```
Comment (arsenm) [inline llvm/lib/Support/YAMLTraits.cpp:803]: Can you use a StringRef with take_back incrementing? (I guess this is just moving code and shouldn't change here)

File: llvm/test/Transforms/WholeProgramDevirt/import-indir.ll [hunk 3]
```diff
@@ -3,7 +3,7 @@
 ; RUN: FileCheck --check-prefix=SUMMARY %s < %t
 
 ; SUMMARY:     GlobalValueMap:
-; SUMMARY-NEXT:  42:
+; SUMMARY-NEXT:  '42':
```
Comment (arsenm) [inline llvm/test/Transforms/WholeProgramDevirt/import-indir.ll:6]: Should quotes only be emitted in the necessary cases?

**Review by arsenm — COMMENTED** [review]
> here are also cases where quoting
isn't strictly necessary, but prevents issues in some cases, for instance strings
consisting of digits only.

What is the issue with just digit only keys?

**Comment by jasilvanus** [comment]
> What is the issue with just digit only keys?

I re-used the existing `needsQuotes` function that is already used for string values implemented here which quotes numeric strings:

https://github.com/llvm/llvm-project/blob/a02019960b1a693320cd43b0ed6653d95877b94f/llvm/include/llvm/Support/YAMLTraits.h#L674

I originally thought there was some potential issue with numeric keys, but now I'm quite confident this just
to preserve the string type, in order to be able to differentiate the string value `'42'` from the numeric value `42`.

For string values, this totally makes sense as we have typed value setters, so if one wants the unquoted variant one should not set a string value to begin with.
However, keys currently always are strings, and thus there is no way to differentiate the keys `'42'` and `42`.

In the last commit, I extended `needsQuotes` to add a flag whether preservation of string values is required, and not require quoting in the `isNull`/`isBool`/`isNumeric` cases then.

File: llvm/lib/Support/YAMLTraits.cpp (hunk 2 above)
Comment (jasilvanus) [reply llvm/lib/Support/YAMLTraits.cpp:804]: I'd prefer to keep it as-is as this code is just moved.

**Review by arsenm — APPROVED** [review]

**Comment by jasilvanus** [comment]
I couldn't repro the test failure locally, trying to merge `main` again and re-trigger CI.

**Pull request merged**